Transfer ownership

juliushaertl opened the pull request:
Rebased version of https://github.com/nextcloud/deck/pull/1955

closes #1306 

## ToDo:

- [x] Ensure that unit tests pass
- [x] Make sure that psalm static code analysis passes
- [x] Add integration tests for transferring a board with example data to a different user
- [x] Follow up PR: UI
  - [x] Create API endpoint to trigger the transfer logic 
  - [x] Add option to the UI to transfer a board to a different user

stefan-niedermann commented:
Will there also be some kind of REST API?

juliushaertl commented:
As of now this is only an occ command for admins. Whenever we expose it as a feature to users we will add an API for the user interface that can be used for the apps as well then.

juliushaertl commented:
@matchish I took the liberty to rebase and adjust some small issues i found during testing. Would be great if you could have another look if all that makes sense to you.

The major change is that when transfering ACL rules it should be enough to just drop the one for the new owner. The old owner will no longer have access unless there is a new rule added manually afterwards.

estan commented:
It would be great if some dev could find time to review this. We have employees leaving and naturally want to transfer the ownership of some of their boards to other members of the team. We rather not mess about with running manual SQL statements against our database, even if it is probably safe if we get it right.

max-nextcloud commented:
I rebased this, will look into the failing tests and review the code changes next.

juliushaertl commented on lib/Db/AclMapper.php:
```diff
@@ -57,4 +57,25 @@ public function findByParticipant($type, $participant): array {
+	 * @param $newOwnerId
+	 * @return void
+	 */
+	public function transferOwnership($boardId, $ownerId, $newOwnerId) {
```
Renamed to deleteParticipantFromBoard and moved to use the query builder.

juliushaertl commented on lib/Db/AssignmentMapper.php:
```diff
@@ -146,4 +146,28 @@ private function getOrigin(Assignment $assignment) {
+			'type' => AssignedUsers::TYPE_USER
+		];
+		$qb = $this->db->getQueryBuilder();
+		$sql = "DELETE FROM `*PREFIX*{$this->tableName}`  WHERE `participant` = :newOwner AND `type`= :type";
```
I adapted the logic quite a bit here in a fresh function called remapAssignedUser to process all cards of a given board id. Overall this will end up in having multiple queries per transferred board but processing all user boards that way now also allows to give intermediate status update on the transfer during the running operation.

juliushaertl commented:
@max-nextcloud I'd appreciate another review. I addressed the remaining comments and also added a dedicated test case for the bug you encountered about the too wide scope of changing assigned users and card ownership: https://github.com/nextcloud/deck/pull/2496/files#diff-aafe9743edf16cdb498c7c4f15245d4d7dd70b1567d28a2b329a709cb950aa03R260-R278

juliushaertl commented:
In addition now the logic of transferring all user boards was changed a bit to just iterate over all boards and transfer them individually, which makes code more sharable and the performance impact of individual queries is ok form my perspective for a one time action of the transfer.

There is also a command line option now if the assigned users and creator of the cards should actually be changed or remain unchanged as well.

juliushaertl commented on lib/Controller/BoardController.php:
```diff
@@ -155,4 +155,15 @@ public function deleteAcl($aclId) {
+	 * * @return null|void
+	 */
+	public function transferOwner($boardId, $owner, $newOwner) {
+		return $this->boardService->transferOwnership($owner, $newOwner);
```
```suggestion
		return $this->boardService->transferBoardOwnership($boardId, $owner, $newOwner);
```

To make use of the newly added method to only transfer a single board.

juliushaertl commented on lib/Controller/BoardController.php:
```diff
@@ -155,4 +155,15 @@ public function deleteAcl($aclId) {
+	 * @param $newOwner
+	 * * @return null|void
+	 */
+	public function transferOwner($boardId, $owner, $newOwner) {
```
```suggestion
	public function transferOwner(int $boardId, string $newOwner): DataResponse {
```

Things to add here:
- a permission check if the current user is actually the owner
- then the owner does not need to be passed as a parameter anymore through the API

juliushaertl commented on lib/Service/BoardService.php:
```diff
@@ -746,6 +746,12 @@ private function clearBoardsCache() {
 		$this->boardsCache = null;
 	}
 
+	private function getBoardOwner($boardId) {
```
Seems to be unused

juliushaertl commented on src/components/board/SharingTabSidebar.vue:
```diff
@@ -53,6 +53,9 @@
 					<ActionCheckbox v-if="canManage" :checked="acl.permissionManage" @change="clickManageAcl(acl)">
 						{{ t('deck', 'Can manage') }}
 					</ActionCheckbox>
+					<ActionCheckbox v-if="canManage" :checked="acl.owner" @change="clickTranserOwner(acl.participant.uid)">
```
```suggestion
					<ActionCheckbox v-if="isCurrentUser(board.owner.uid)" :checked="acl.owner" @change="clickTranserOwner(acl.participant.uid)">
```

juliushaertl reviewed (commented):
Few additional comments, but looks nice :)

max-nextcloud reviewed (changes requested):
only checked the code so far. Saw other comments by @juliushaertl so leaving it to that for now.

luka-nextcloud commented:
@max-nextcloud @juliushaertl Please review again.

kabunatu reviewed (approved):
Good to go from my side 👍

juliushaertl reviewed (commented):
👍 Good to go from my side

max-nextcloud reviewed (approved):
Looks good to me. Adding one comment about uncaught exceptions. I have not tried these corner cases myself but i suspect they result in error traces being printed rather than a proper error message - which can be confusing for users.

vinicius73 commented on lib/Controller/BoardController.php:
```diff
@@ -150,9 +153,24 @@ public function deleteAcl($aclId) {
+	 * @param $boardId
+	 * @param $owner
+	 * @param $newOwner
+	 * * @return null|void
```
```suggestion
	 * @return DataResponse
```

juliushaertl commented on composer.json:
```diff
@@ -14,7 +14,7 @@
 	"require-dev": {
 		"roave/security-advisories": "dev-master",
-		"christophwurst/nextcloud": "^21@dev",
+		"christophwurst/nextcloud": "dev-master",
```
Yes, we should test master against master of server, so not related to the PR but should be that way and it solved some psalm issues in this PR.

juliushaertl commented:
/backport to stable23

juliushaertl commented:
/backport to stable22
